### PR TITLE
PREAPPS-6109: Implement saveDraft and SendMessage flow for Offline Desktop

### PR DESCRIPTION
**Pull request by lavegupta**

### CODE REVIEW

File: src/batch-client/index.ts [hunk 1]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
 			return ids ? ids.split(',') : [];
 		});
 
+	// For offline Drafts
+	public attach = (files: any, { ...message }) => {
+		const promises = castArray(files).map(file => this.localStoreClient.attach({ file }));
+
+		return Promise.all(promises).then(attached => {
```
Comment (madhavdhuppe) [inline src/batch-client/index.ts:443]: can we .catch()

File: src/batch-client/index.ts [hunk 2]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+			// After all uploads have completed, add attachments to the message and save it as a draft.
+			attached.forEach((attachment, index) => {
+				if (attachment) {
+					const file = files[index];
+					if (attachment.contentDisposition === 'inline') {
+						const { base64, url, part, messageId } = attachment;
+						const contentType = file.type || file.contentType;
```
Comment (madhavdhuppe) [inline src/batch-client/index.ts:450]: file destructring

File: src/batch-client/index.ts [hunk 3]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
 
+	// For offline Drafts
+	public attach = (files: any, { ...message }) => {
+		const promises = castArray(files).map(file => this.localStoreClient.attach({ file }));
+
+		return Promise.all(promises).then(attached => {
+			// After all uploads have completed, add attachments to the message and save it as a draft.
+			attached.forEach((attachment, index) => {
+				if (attachment) {
+					const file = files[index];
```
Comment (madhavdhuppe) [inline src/batch-client/index.ts:447]: i think we should move const file = files[index]; inside inline if

File: src/batch-client/index.ts [hunk 4]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
 			return ids ? ids.split(',') : [];
 		});
 
+	// For offline Drafts
+	public attach = (files: any, { ...message }) => {
```
Comment (silentsakky) [inline src/batch-client/index.ts:440]:
```suggestion
	public attach = (files: any, message) => {
```

Comment (silentsakky) [reply src/batch-client/index.ts:440]: is this something we can move to local client class? don't want to put local client related code to api-client

File: src/batch-client/index.ts [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import DataLoader from 'dataloader';
 import castArray from 'lodash/castArray';
+import differenceWith from 'lodash/differenceWith';
```
Comment (silentsakky) [inline src/batch-client/index.ts:3]: update rollup.config.js to specify this as external method

File: src/batch-client/index.ts [hunk 6]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+							contentType,
+							size: file.size,
+							contentId: file.contentId,
+							contentDisposition: file.contentDisposition,
+							...(part && { part }),
+							...(messageId && { messageId }),
+							...(base64 && { base64 }),
+							...(url
+								? { url }
+								: base64 && contentType && { url: `data:${contentType};base64,${base64}` })
```
Comment (silentsakky) [inline src/batch-client/index.ts:462]: in which case we will get base64 data?

File: src/batch-client/index.ts [hunk 7]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+						const fileAttributes = {
+							filename: file.name || file.filename,
+							contentType,
+							size: file.size,
+							contentId: file.contentId,
+							contentDisposition: file.contentDisposition,
+							...(part && { part }),
+							...(messageId && { messageId }),
+							...(base64 && { base64 }),
+							...(url
```
Comment (silentsakky) [inline src/batch-client/index.ts:460]: looks like we are merging properties from file and attachment object, can we merge using destructure and then update some properties?

File: src/batch-client/index.ts [hunk 8]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+								: base64 && contentType && { url: `data:${contentType};base64,${base64}` })
+						};
+						message.inlineAttachments = [
+							...(message.inlineAttachments || []).map(({ __typename, ...rest }: any) => ({
+								...rest
+							})),
+							fileAttributes
+						];
+					} else {
+						message.attachments = [...(message.attachments || []), {
```
Comment (silentsakky) [inline src/batch-client/index.ts:471]: easier to use concat for array concatenation instead of destructure

File: src/batch-client/index.ts [hunk 9]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+							...attachment,
+							part: `${(message.attachments || []).length + 2}`
```
Comment (silentsakky) [inline src/batch-client/index.ts:473]: + 2 ???

File: src/batch-client/index.ts [hunk 10]
```diff
@@ -433,6 +436,57 @@ export class ZimbraBatchClient {
...
+						}];
+					}
+				}
+			});
+
+			message.uploadingFileList = differenceWith(
```
Comment (silentsakky) [inline src/batch-client/index.ts:479]: this specific property is used in zm-x-web only, it's not linked to server implementation, i doubt we should be having this here

File: src/batch-client/index.ts (hunk 5 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Removed

File: src/batch-client/index.ts (hunk 6 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Code moved to zm-x-web

File: src/batch-client/index.ts (hunk 7 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Code refactored and moved to zm-x-web

File: src/batch-client/index.ts (hunk 8 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Code refactored and moved to zm-x-web

File: src/batch-client/index.ts (hunk 4 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Done and moved to zm-x-web

File: src/batch-client/index.ts (hunk 3 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Code refactored and moved to zm-x-web

File: src/batch-client/index.ts (hunk 2 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Code refactored and moved to zm-x-web

File: src/batch-client/index.ts (hunk 1 above)
Comment (lavegupta) [reply src/batch-client/index.ts]: Please review https://github.com/ZimbraOS/zm-x-web/pull/3473

**Review by silentsakky — APPROVED** [review]

**Review by sachinpatel88 — APPROVED** [review]

**Review by sachinpatel88 — APPROVED** [review]

**Pull request merged**